resolver/scopes.go

```diff
@@ -116,99 +107,70 @@ func domainInScope(dotPrefixedFQDN string, scopeList []string) bool {
 }
 
 // GetResolversInScope returns all resolvers that are in scope the resolve the given query and options.
-func GetResolversInScope(ctx context.Context, q *Query) (selected []*Resolver) { //nolint:gocognit // TODO
+func GetResolversInScope(ctx context.Context, q *Query) (selected []*Resolver, tryAll bool) { //nolint:gocognit // TODO
 	resolversLock.RLock()
 	defer resolversLock.RUnlock()
 
-	// resolver selection:
-	// local -> local scopes, mdns
-	// local-inaddr -> local, mdns
-	// global -> local scopes, global
-	// special -> local scopes, local
+	// Internal use domains
+	if strings.HasSuffix(q.dotPrefixedFQDN, internalSpecialUseDomainScope) {
+		return envResolvers, false
+	}
 
-	// special connectivity domains
+	// Special connectivity domains
 	if netenv.IsConnectivityDomain(q.FQDN) && len(systemResolvers) > 0 {
-		selected = append(selected, envResolver)
+		// Do not do compliance checks for connectivity domains.
 		selected = append(selected, systemResolvers...) // dhcp assigned resolvers
-		return selected
+		return selected, false
 	}
 
-	// check local scopes
+	// Prioritize search scopes
 	for _, scope := range localScopes {
 		if strings.HasSuffix(q.dotPrefixedFQDN, scope.Domain) {
-			// scoped resolvers
-			for _, resolver := range scope.Resolvers {
-				if err := resolver.checkCompliance(ctx, q); err == nil {
-					selected = append(selected, resolver)
-				} else {
-					log.Tracef("skipping non-compliant resolver: %s", resolver.Server)
-				}
-			}
+			selected = addResolvers(ctx, q, selected, scope.Resolvers)
 		}
 	}
-	// if there was a match with a local scope, stop here
-	if len(selected) > 0 {
-		// add mdns
-		if err := mDNSResolver.checkCompliance(ctx, q); err == nil {
-			selected = append(selected, mDNSResolver)
-		} else {
-			log.Tracef("skipping non-compliant resolver: %s", mDNSResolver.Server)
+
+	// Handle multicast domains
+	if domainInScope(q.dotPrefixedFQDN, multicastDomains) {
+		selected = addResolvers(ctx, q, selected, mDNSResolvers)
+		// Add local resolvers if no resolvers were selected.
+		if len(selected) == 0 {
+			selected = addResolvers(ctx, q, selected, localResolvers)
 		}
```

Comment:
why only add them if there are no `mDNSResolvers`. IMHO it should either always query them or not.